File: ruby/logbrew-ruby/lib/logbrew/telemetry_resource.rb
# frozen_string_literal: true

module LogBrew
  # Immutable service, deployment, runtime, framework, OS, device, and app identity.
  class TelemetryResource
    SECTION_FIELDS = {
      "service" => %w[name version].freeze,
      "deployment" => %w[environment release].freeze,
      "runtime" => %w[name version].freeze,
      "framework" => %w[name version].freeze,
      "operatingSystem" => %w[name version build].freeze,
      "device" => %w[family model architecture].freeze,
      "application" => %w[name version build].freeze
    }.freeze
    NAME_REQUIRED_SECTIONS = %w[service runtime framework operatingSystem].freeze

    def self.create
      TelemetryResourceBuilder.new
    end

    def self.from_hash(value = nil, **keywords)
      if value.nil?
        value = keywords
      elsif !keywords.empty?
        raise TelemetryContextValue.invalid("telemetry resource must be one object")
      end
      object = TelemetryContextValue.object(value, "telemetry resource")
      TelemetryContextValue.reject_unknown_fields(object, SECTION_FIELDS.keys, "telemetry resource")
      raise TelemetryContextValue.invalid("telemetry resource must not be empty") if object.empty?

      normalized = {}
      SECTION_FIELDS.each do |section_name, fields|
        next unless object.key?(section_name)

        section = TelemetryContextValue.object(
          object.fetch(section_name),
          "telemetry resource #{section_name}"
        )
        TelemetryContextValue.reject_unknown_fields(
          section,
          fields,
          "telemetry resource #{section_name}"
        )
        normalized_section = {}
        fields.each do |field|
          next unless section.key?(field)

          normalized_section[field] = TelemetryContextValue.required_string(
            section.fetch(field),
            "telemetry resource #{section_name} #{field}"
          )
        end
        if NAME_REQUIRED_SECTIONS.include?(section_name) && !normalized_section.key?("name")
          raise TelemetryContextValue.invalid("telemetry resource #{section_name} name is required")
        end
        if normalized_section.empty?
          raise TelemetryContextValue.invalid("telemetry resource #{section_name} must not be empty")
        end
        normalized[section_name] = normalized_section
      end

      new(normalized)
    end

    def self.merge(base, override)
      require_resource_or_nil(base, "base telemetry resource")
      require_resource_or_nil(override, "override telemetry resource")
      return nil if base.nil? && override.nil?
      return from_hash(override.to_h) if base.nil?
      return from_hash(base.to_h) if override.nil?

      base_value = base.to_h
      override_value = override.to_h
      merged = {}
      SECTION_FIELDS.each_key do |section|
        base_section = base_value[section]
        override_section = override_value[section]
        next if base_section.nil? && override_section.nil?

        merged[section] = (base_section || {}).merge(override_section || {})
      end
      from_hash(merged)
    end

    def to_h
      TelemetryContextValue.deep_copy(@value)
    end

    private

    def initialize(value)
      @value = TelemetryContextValue.deep_freeze(TelemetryContextValue.deep_copy(value))
      freeze
    end

    def self.require_resource_or_nil(value, label)
      return if value.nil? || value.is_a?(TelemetryResource)

      raise TelemetryContextValue.invalid("#{label} must be a LogBrew::TelemetryResource")
    end
    private_class_method :require_resource_or_nil
  end

  # Builder for one immutable telemetry resource.
  class TelemetryResourceBuilder
    def initialize
      @value = {}
    end

    def with_service(name:, version: nil)
      @value["service"] = named_version(name, version)
      self
    end

    def with_deployment(environment: nil, release: nil)
      @value["deployment"] = optional_fields(environment: environment, release: release)
      self
    end

    def with_runtime(name:, version: nil)
      @value["runtime"] = named_version(name, version)
      self
    end

    def with_framework(name:, version: nil)
      @value["framework"] = named_version(name, version)
      self
    end

    def with_operating_system(name:, version: nil, build: nil)
      @value["operatingSystem"] = optional_fields(name: name, version: version, build: build)
      self
    end

    def with_device(family: nil, model: nil, architecture: nil)
      @value["device"] = optional_fields(family: family, model: model, architecture: architecture)
      self
    end

    def with_application(name: nil, version: nil, build: nil)
      @value["application"] = optional_fields(name: name, version: version, build: build)
      self
    end

    def build
      TelemetryResource.from_hash(@value)
    end

    private

    def named_version(name, version)
      optional_fields(name: name, version: version)
    end

    def optional_fields(**values)
      values.each_with_object({}) do |(key, value), section|
        section[key.to_s] = value unless value.nil?
      end
    end
  end
end

File: php/logbrew-php/src/TelemetryResource.php
<?php

declare(strict_types=1);

namespace LogBrew;

/** Immutable bounded service, deployment, runtime, framework, OS, device, and application identity. */
final class TelemetryResource
{
    /** @param array<string, array<string, string>> $value */
    private function __construct(private readonly array $value)
    {
    }

    /** Return a builder for one non-empty telemetry resource. */
    public static function create(): TelemetryResourceBuilder
    {
        return new TelemetryResourceBuilder();
    }

    /**
     * Validate and detach a schema-shaped resource array.
     *
     * @param array<mixed> $value
     */
    public static function fromArray(array $value): self
    {
        $value = TelemetryContextValue::object($value, 'telemetry resource');
        TelemetryContextValue::rejectUnknownFields($value, [
            'service',
            'deployment',
            'runtime',
            'framework',
            'operatingSystem',
            'device',
            'application',
        ], 'telemetry resource');
        if ($value === []) {
            throw TelemetryContextValue::invalid('telemetry resource must not be empty');
        }

        $normalized = [];
        foreach ([
            'service' => ['name', 'version'],
            'deployment' => ['environment', 'release'],
            'runtime' => ['name', 'version'],
            'framework' => ['name', 'version'],
            'operatingSystem' => ['name', 'version', 'build'],
            'device' => ['family', 'model', 'architecture'],
            'application' => ['name', 'version', 'build'],
        ] as $sectionName => $fields) {
            if (!array_key_exists($sectionName, $value)) {
                continue;
            }
            $section = TelemetryContextValue::object(
                $value[$sectionName],
                "telemetry resource {$sectionName}"
            );
            TelemetryContextValue::rejectUnknownFields(
                $section,
                $fields,
                "telemetry resource {$sectionName}"
            );
            $normalizedSection = [];
            foreach ($fields as $field) {
                if (!array_key_exists($field, $section) || !is_string($section[$field])) {
                    if (array_key_exists($field, $section)) {
                        throw TelemetryContextValue::invalid(
                            "telemetry resource {$sectionName} {$field} must be a string"
                        );
                    }
                    continue;
                }
                $normalizedSection[$field] = TelemetryContextValue::requiredString(
                    $section[$field],
                    "telemetry resource {$sectionName} {$field}"
                );
            }
            if (in_array($sectionName, ['service', 'runtime', 'framework', 'operatingSystem'], true)
                && !array_key_exists('name', $normalizedSection)) {
                throw TelemetryContextValue::invalid("telemetry resource {$sectionName} name is required");
            }
            if ($normalizedSection === []) {
                throw TelemetryContextValue::invalid("telemetry resource {$sectionName} must not be empty");
            }
            $normalized[$sectionName] = $normalizedSection;
        }

        return new self($normalized);
    }

    /** Field-wise merge with later resource values taking precedence. */
    public static function merge(?self $base, ?self $override): ?self
    {
        if ($base === null) {
            return $override;
        }
        if ($override === null) {
            return $base;
        }

        $merged = [];
        foreach ([
            'service',
            'deployment',
            'runtime',
            'framework',
            'operatingSystem',
            'device',
            'application',
        ] as $section) {
            $baseSection = $base->value[$section] ?? null;
            $overrideSection = $override->value[$section] ?? null;
            if ($baseSection !== null || $overrideSection !== null) {
                $merged[$section] = array_merge($baseSection ?? [], $overrideSection ?? []);
            }
        }

        return self::fromArray($merged);
    }

    /** @return array<string, array<string, string>> */
    public function toArray(): array
    {
        return $this->value;
    }
}
